cli/src/bin.ts

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env node
+
+import { Command } from "commander";
+import * as path from "path";
+import fg from "fast-glob";
+import { watch } from "chokidar";
+import * as fs from "fs";
+import { promisify } from "util";
+import { parseMachinesFromFile } from "xstate-parser-demo";
+import {
+  doesTsTypesRequireUpdate,
+  makeXStateUpdateEvent,
+  writeToTypegenFile,
+} from "xstate-vscode-shared";
+
+const readFile = promisify(fs.readFile);
+const writeFile = promisify(fs.writeFile);
+
+type FileEdit = {
+  start: number;
+  end: number;
+  newText: string;
+};
+
+const processFileEdits = (oldText: string, fileEdits: FileEdit[]): string => {
+  let newText = oldText;
+  for (let i = 0; i < fileEdits.length; i++) {
+    const fileEdit = fileEdits[i];
+    const futureFileEdits = fileEdits.slice(i + 1);
+    newText =
+      newText.slice(0, fileEdit.start) +
+      fileEdit.newText +
+      newText.slice(fileEdit.end);
+
+    /**
+     * If the file edits are after this one, increment them
+     * by the length of the new text MINUS the length of the old
+     */
+    const oldTextLength = fileEdit.end - fileEdit.start;
+    const lengthDifference = fileEdit.newText.length - oldTextLength;
+    futureFileEdits.forEach((futureFileEdit) => {
+      if (futureFileEdit.start > fileEdit.end) {
+        futureFileEdit.start += lengthDifference;
+        futureFileEdit.end += lengthDifference;
+      }
+    });
```

Comment:
Perhaps instead of this, you could sort file edits in the reversed order and just apply them from the last one to the first one. This potentially assumes that text edits can't be overlapping but I think this is sort of an invariant of this anyway - I don't see a situation when we'd like to create overlapping text edits

Reply:
Urgh, that's so obvious and so much simpler. Great idea!